### ISSUE TITLE
 #57 related field

### ISSUE THREAD

**Issue by sakulstra** [vazco/uniforms#296]
Hello,
i got two select fields where the input of one is related to(acts as filter for) the seconds one:
```js
// options
options1: [
 "BRANDS", "COLORS"
]

options2: [
{key: "BRANDS", values: ["fritz","miomio","maya"]},
{key: "COLORS", values: ["red","green"]}
]

// select component
const Select = ({
  allowedValues,
  disabled,
  name,
  onChange,
  value,
  filter,
  findValue,
  findField,
  relatedField,
   ...props
}, {uniforms: {model}}) => {
// 1. idea using context
console.log(model.topic) // apparently doesn't rerender because context change doesn't trigger a rerender?
// 2. idea using findField
console.log(name, findField('topic')) // similar problem - doesn't rerender as findField doesn't change - only it's return value
  return (
    <Select
      {...props}
      value={value}
      name={name}
      disabled={disabled}
      options={allowedValues}
      valueKey={valueKey}
      labelKey={labelKey}
      onChange={item => onChange(item)}
    />
  )
}

Select.contextTypes = BaseField.contextTypes;
export default connectField(Select)

// 3. idea using 
<AutoForm
              schema={sessionSchema}
              onSubmit={this.addSession}
              ref={ref => formRef = ref}
            >
<Select options={formRef && options.filter(value => formRef.getModel() === somecoolfilter)} />
// formRef seems to be undefined in this context
<AutoForm/>
```

What's the proper way to archive that?

---
related to #57

**Comment by radekmie** [comment]
There are at least few ways to achieve that, so here are three ideas:

First one is limited to nested fields only _(both filter and filtered need to be nested under one key)_: simply create a field for _upper_ field and manage props as you want.

Second one is not very performant _(it's only a proof of concept - it might be optimized!)_:

```js
const NoCoolNameSoFar = (props, context) =>
  React.cloneElement(props.children, props.fromContext(context.uniforms))
;

NoCoolNameSoFar.contextTypes = BaseField.contextTypes;

// Usage:
const schema = new SimpleSchema({
    filter: {
        type: String,
        allowedValues: ['cats', 'dogs'],
        defaultValue: 'cats'
    },
    filtered: {
        type: String,
        uniforms: {
            filterFrom: 'filter',
            filterWith: {
                cats: ['cat 1', 'cat 2'],
                dogs: ['dog 1', 'dog 2']
            }
        }
    }
});

<AutoForm schema={schema}>
  <SelectField name="filter" />
  <NoCoolNameSoFar fromContext={uniforms => {
    const {filterFrom, filterWith} = uniforms.schema.getProps('filtered');

    // TODO: filterFrom may be nested
    const value = uniforms.model[filterFrom];

    return {allowedValues: filterWith[value] || []};
  }}>
    <SelectField name="filtered" />
  </NoCoolNameSoFar>
</AutoForm>
```

Third one _(not performant, but the easiest)_:
```js
// Let's assume, that X is using findValue or one of its friends.
class XField extends connectField(X) {
    shouldComponentUpdate () {
        return true;
    }
}
```

As I've looked into the code, it's not so easy right now to detect a change of other field in another one. Immutable state might be helpful here, but it's still a distant future. For now, use one of these three above, or find out another way - there are plenty of them, I'm sure.

Let me know, if these are working for you - I've tested them shallowly.

**Comment by sakulstra** [comment]
@radekmie thank you for that awesome response. Yeah actually I have a 4th way working right now which sees super complicated, so i'll try one of yours.

just to mention it:
currently i'm using an addition observable mobx store which i populate onChange in addition to the uniforms model and automatically rerenders the observing componet

**Comment by radekmie** [comment]
Sometimes I'm surprised, what solutions people might think out, when in hurry :D Anyway - if you think that's all, please, close this issue then.

**Comment by sakulstra** [comment]
@radekmie yeah we're already had a mobx store in place so it was basically just two lines to make it handle the updates. But i see you're point - it's was a pretty dirty solution xD

Just for completeness here's a i think not too dirty solution based on you're third suggestion.
The idea is, that the select should now what it is related to beforehand so I added a `relatedField` prop.
We now only update when 
1.`relatedField` prop is set and the `model[relatedField]` was changed
2. when uniform would have updated anyways

// simplified code as we handle some other cases like multiple related fields
```
export class HackyMultiSelect extends connectField(MultiSelect) {
  shouldComponentUpdate (p, s, nConext) {
    if (p.relatedField &&
      nConext.uniforms.model[p.relatedField] !== this.context.uniforms.model[p.relatedField]) return true
    return super.shouldComponentUpdate(p, s, nConext)
  }
}
```
